### PR TITLE
Adding section on using building blocks to express inheritance idioms

### PR DESCRIPTION
**Pull request by gusthoff**
Based on document written by Patrick Rogers.

### CODE REVIEW

File: content/courses/ada-idioms/chapters/inheritance_idioms.rst [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+There are two *solutions* in this entry, one for each of the two inheritance
+idioms under discussion. First, we will specify our building block choices,
+then we will show the two idiom expressions in separate subsections.
+
+- We will use tagged types for the sake of the full OOP capabilities. That is
+  the most common choice when inheritance is involved. The static form of
+  inheritance has cases in which it is useful, however those cases are very
+  narrow in applicability.
+
+- We will assume that both the parent type and the child type are both private
```
Comment (pat-rogers) [inline content/courses/ada-idioms/chapters/inheritance_idioms.rst:191]: "We will assume that both the parent" should be "We will assume that the parent"

File: content/courses/ada-idioms/chapters/inheritance_idioms.rst [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+- We are going to declare the child type in a distinct, dedicated package,
+  following the :ref:`ADT idiom <Ada_Idioms_Abstract_Data_Types>`. This package
+  may or may not be a hierarchical child of the parent package. The solution
+  approach does not require a child package's special compile-time visibility,
+  although a child package is often necessary for the sake of that visibility.
+
+- Whether the child type is visibly derived will vary with the
+  :ref:`inheritance idiom <Ada_Idioms_Implementation_Inheritance>` solution.
+
+To avoid unnecessary code duplication, the two idiom solutions use the same
```
Comment (pat-rogers) [inline content/courses/ada-idioms/chapters/inheritance_idioms.rst:208]: "duplication, the two idiom solutions" should be "duplication, examples for the two idiom solutions"

File: content/courses/ada-idioms/chapters/inheritance_idioms.rst [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+parent. Hence we declare the private type as a private extension in the visible
+part of the package:
+
+.. code-block:: ada
+
+    with P; use P;
+    package Q is
+      type Child is new Parent with private;
+      --  implicit, inherited primitive Parent operations declared here,
+      --  now for type Child additional primitives for Child explicitly
```
Comment (pat-rogers) [inline content/courses/ada-idioms/chapters/inheritance_idioms.rst:242]: "primitives for Child explicitly" should start a new comment on its own line

File: content/courses/ada-idioms/chapters/inheritance_idioms.rst [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+      --  declared, if any
+    private
+      type Child is new Parent with record ... end record;
+    end Q;
+
+The primitive operations from the parent type are implicitly, automatically
+declared immediately after the private extension declaration. That means those
+operations are in the visible part of the package, hence clients can invoke
+them. Any additional operations for the interface will be explicitly declared in
```
Comment (pat-rogers) [inline content/courses/ada-idioms/chapters/inheritance_idioms.rst:251]: "for the interface" should be "for the client interface"

File: content/courses/ada-idioms/chapters/inheritance_idioms.rst [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+.. _Ada_Idioms_Implementation_Inheritance:
+
+Implementation Inheritance
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Recall that with Implementation Inheritance clients do not have compile-time
+visibility to the *is-a* relationship between the parent and child types. We
+meet that requirement by not making the child visibly derived from the parent.
+Therefore, we declare the child type as a simple tagged private type and only
+mention the parent in the child type's completion located in the package private
```
Comment (pat-rogers) [inline content/courses/ada-idioms/chapters/inheritance_idioms.rst:439]: Let's strike the word "located"

File: content/courses/ada-idioms/chapters/inheritance_idioms.rst (hunk 3 above)
Comment (gusthoff) [reply content/courses/ada-idioms/chapters/inheritance_idioms.rst:242]: Ok, I've just added a newline before the word "additional".

**Review by pat-rogers — APPROVED** [review]
Thanks!

**Pull request merged**